components/backup-stream/src/checkpoint_manager.rs

```diff
@@ -76,8 +159,81 @@ impl CheckpointManager {
         self.items.clear();
     }
 
+    pub fn spawn_subscription_mgr(&mut self) -> future![()] {
+        let (tx, rx) = async_mpsc::channel(1024);
+        let sub = SubscriptionManager {
+            subscribers: Default::default(),
+            input: rx,
+        };
+        self.manager_handle = Some(tx);
+        sub.main_loop()
+    }
+
+    pub fn update_region_checkpoints(&mut self, region_and_checkpoint: Vec<(Region, TimeStamp)>) {
+        for (region, checkpoint) in &region_and_checkpoint {
+            self.do_update(region, *checkpoint);
+        }
+
+        self.notify(region_and_checkpoint.into_iter());
+    }
+
     /// update a region checkpoint in need.
+    #[cfg(test)]
     pub fn update_region_checkpoint(&mut self, region: &Region, checkpoint: TimeStamp) {
+        self.do_update(region, checkpoint);
+        self.notify(std::iter::once((region.clone(), checkpoint)));
+    }
+
+    pub fn add_subscriber(&mut self, sub: Subscription) -> future![Result<()>] {
+        let mgr = self.manager_handle.as_ref().cloned();
+
+        // NOTE: we cannot send the real error into the client directly because once
+        // we send the subscription into the sink, we cannot fetch it again :(
+        async move {
+            let mgr = mgr.ok_or(Error::Other(box_err!("subscription manager not get ready")));
+            let mut mgr = match mgr {
+                Ok(mgr) => mgr,
+                Err(err) => {
+                    sub.fail(RpcStatus::with_message(
+                        RpcStatusCode::UNAVAILABLE,
+                        "subscription manager not get ready.".to_owned(),
+                    ))
+                    .await
+                    .map_err(|err| {
+                        annotate!(err, "failed to send request to subscriber manager")
+                    })?;
+                    return Err(err);
+                }
+            };
+            mgr.send(SubscriptionOp::Add(sub))
+                .await
+                .map_err(|err| annotate!(err, "failed to send request to subscriber manager"))?;
+            Ok(())
+        }
+    }
+
+    fn notify(&mut self, items: impl Iterator<Item = (Region, TimeStamp)>) {
+        if let Some(mgr) = self.manager_handle.as_mut() {
+            let r = items
+                .map(|(r, ts)| {
+                    let mut f = FlushEvent::new();
+                    f.set_checkpoint(ts.into_inner());
+                    f.set_start_key(r.start_key);
+                    f.set_end_key(r.end_key);
+                    f
+                })
+                .collect::<Box<[_]>>();
```

Comment:
Why not `collect::<Vec<_>>`?

Reply:
Because the slice will never be appended, so we collect it as boxed slice to get rid of the `capacity` field in `Vec`.

(Also note that the message `Emit(Box<[FlushEvent]>)` uses boxed slice too.)